File: app/src/main/java/mydebts/android/app/feature/events/EventsFragment.java
package mydebts.android.app.feature.events;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import javax.inject.Inject;

import mydebts.android.app.R;
import mydebts.android.app.data.EventsSource;
import mydebts.android.app.data.model.Event;
import mydebts.android.app.di.SubcomponentBuilderResolver;
import mydebts.android.app.feature.main.MainRouter;
import mydebts.android.app.rx.RxUtil;

public class EventsFragment extends Fragment {

    @Inject EventsSource eventsSource;
    @Inject RxUtil rxUtil;

    RecyclerView eventsRecyclerView;
    View emptyView;

    EventsAdapter adapter;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);

        ((EventsSubcomponent.Builder) SubcomponentBuilderResolver.resolve(this))
                .build().inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.activity_events, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        ViewBinder.bind(this);

        eventsSource.getAll()
                .compose(rxUtil.singleSchedulersTransformer())
                .subscribe(this::setEvents, this::setError);
    }

    void onAddEventClick() {
        ((MainRouter)getActivity()).navigateToNewEvent();
    }

    void onEventClick(Event event) {
        ((MainRouter)getActivity()).navigateToEvent(event);
    }

    private void setEvents(List<Event> events) {
        boolean isEmpty = events == null || events.size() == 0;

        emptyView.setVisibility(isEmpty ? View.VISIBLE : View.GONE);
        eventsRecyclerView.setVisibility(isEmpty ? View.GONE : View.VISIBLE);

        adapter.setEvents(events);
    }

    private void setError(Throwable throwable) {

    }
}

File: app/src/main/java/mydebts/android/app/feature/events/EventsFragment.kt
package mydebts.android.app.feature.events

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.RecyclerView
import android.view.ActionMode
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Menu
import android.view.MenuItem
import android.view.MenuInflater
import dagger.android.support.AndroidSupportInjection

import javax.inject.Inject

import mydebts.android.app.R
import mydebts.android.app.data.model.Event
import mydebts.android.app.feature.event.EventActivity
import mydebts.android.app.feature.people.PeopleActivity

class EventsFragment : Fragment(), ActionMode.Callback, EventsScreen {

    @Inject lateinit var presenter: EventsPresenter

    private lateinit var adapter: EventsAdapter

    private var eventsRecyclerView: RecyclerView? = null
    private var emptyView: View? = null

    private var actionMode: ActionMode? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
        setHasOptionsMenu(true)

        AndroidSupportInjection.inject(this)

        adapter = EventsAdapter(presenter.getEvents(), presenter.getSelections())

        adapter.setOnItemClickListener {
            if (actionMode == null) {
                presenter.onEventClick(it)
            } else {
                presenter.onSelectEvent(it)
            }
        }

        adapter.setOnItemLongClickListener { position ->
            actionMode.takeIf { actionMode == null }?.let {
                presenter.onSelectEvent(position)
                actionMode = activity.startActionMode(this)
                true
            } == true
        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater?.inflate(R.layout.activity_events, container, false)

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        activity.setTitle(R.string.title_events)

        view?.let {
            eventsRecyclerView = view.findViewById(R.id.list_events)
            eventsRecyclerView?.adapter = adapter

            emptyView = view.findViewById(R.id.text_no_events)

            val addEventButton: View = view.findViewById(R.id.button_add_event)
            addEventButton.setOnClickListener { startActivity(EventActivity.newIntent(activity)) }
        }

        presenter.onViewCreated()
    }

    override fun onDestroyView() {
        presenter.onDestroyView()
        super.onDestroyView()
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater?.inflate(R.menu.menu_events, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean = item?.let {
        when (it.itemId) {
            R.id.action_switch_to_people -> {
                startActivity(PeopleActivity.newIntent(activity))
                true
            }
            else -> false
        }
    } == true

    override fun onActionItemClicked(mode: ActionMode?, item: MenuItem?): Boolean = item?.let {
        when (it.itemId) {
            R.id.action_delete -> {
                presenter.deleteSelectedEvents()
                mode?.finish()
                true
            }
            else -> false
        }
    } == true

    override fun onCreateActionMode(mode: ActionMode?, menu: Menu?): Boolean {
        mode?.menuInflater?.inflate(R.menu.context_menu_event, menu)
        return true
    }

    override fun onPrepareActionMode(mode: ActionMode?, menu: Menu?): Boolean = false

    override fun onDestroyActionMode(mode: ActionMode?) {
        actionMode = null
        presenter.clearSelections()
    }

    override fun showEvent(event: Event) {
        startActivity(EventActivity.newIntent(activity, event))
    }

    override fun setEventsVisibility(visibility: Int) {
        eventsRecyclerView?.visibility = visibility
    }

    override fun setEmptyViewVisibility(visibility: Int) {
        emptyView?.visibility = visibility
    }

    override fun notifyDataSetChanged() {
        adapter.notifyDataSetChanged()
    }

    override fun notifyItemChanged(position: Int) {
        adapter.notifyItemChanged(position)
    }

    override fun notifyItemRemoved(position: Int) {
        adapter.notifyItemRemoved(position)
    }
}
